uitools/import/Esri/ArcGISRuntime/Toolkit/OAuth2View.qml

```diff
@@ -49,23 +49,59 @@ Dialog {
             centerIn: parent
             fill: parent
         }
-        
+
         url: controller.currentChallengeUrl
 
         onLoadingChanged: {
             if (loadRequest.status === WebView.LoadSucceededStatus) {
                 forceActiveFocus();
+            } else if (loadRequest.status === WebView.LoadFailedStatus) {
+                controller.cancelWithError("Failed to load");
             }
         }
 
         onTitleChanged: {
-            if (title.indexOf("SUCCESS code=") > -1) {
-                var authCode = title.replace("SUCCESS code=", "");
+            // If the title contains "SUCCESS", get the authorization code from the title and continue authenticationChallenge.
+            if (isSuccess()) {
+                const authCode = title.replace("SUCCESS code=", "");
                 controller.continueWithOAuthAuthorizationCode(authCode);
-            } else if (title.indexOf("Denied error=") > -1) {
-                controller.cancel();
+                return;
             }
+
+            // If the title contains "Denied error=invalid_request", get the HTML content.
+            else if (isInvalidRequest()) {
+                getHtmlContent();
+                return;
+            }
+
+            // If there is an error, cancel with error.
+            if (isError()) {
+                controller.cancelWithError(title, html);
+            }
+        }
+
+        // Property to get HTML content when necessary.
+        property string html: ""
+        onHtmlChanged: {
+            controller.cancelWithError(title, html);
+        }
+
+        // Helper functions
+        function isSuccess() {
+            return title.indexOf("SUCCESS code=") > -1;
+        }
+
+        function isInvalidRequest() {
+            return (title.indexOf("Denied error=invalid_request") > -1);
+        }
+
+        function isError() {
+            return (title.indexOf("Denied error=") > -1) || (title.indexOf("Error: ") > -1);
+        }
+
+        function getHtmlContent() {
```

Comment:
Consider a different name for this helper. Maybe `readHtmlContent`? Since it returns void you aren't really "getting" anything.